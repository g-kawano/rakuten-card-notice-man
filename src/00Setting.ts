export class Setting {
  /**
   * 決済履歴の本人利用部の通知メッセージ表示用名
   */
  DISPLAY_HIMSELF: string | null;

  /**
   * 決済履歴の家族利用部の通知メッセージ表示用名
   */
  DISPLAY_FAMILY: string | null;

  /**
   * 本プロジェクトで使用する Google Drive のフォルダ名
   */
  DRIVE_FOLDER_NAME: string;

  /**
   * LINE Messaging API 用 アクセストークン
   */
  LINE_CHANNEL_ACCESS_TOKEN: string;

  /**
   * LINE トークグループ ID
   */
  LINE_GROUP_ID: string;

  /**
   * マスターデータを登録しているスプレッドシートファイル名
   */
  MASTER_SPREAD_SHEET_FILE: string;

  /**
   * メール受信日
   */
  MESSAGE_DATE: string;

  constructor() {
    const displayHimself = PropertiesService.getScriptProperties().getProperty("DISPLAY_HIMSELF");

    const displayFamily = PropertiesService.getScriptProperties().getProperty("DISPLAY_FAMILY");

    const driveFolderName = PropertiesService.getScriptProperties().getProperty("DRIVE_FOLDER_NAME");
    if (driveFolderName === null) {
      throw new Error("Gas properties DRIVE_FOLDER_NAME is not null!");
    }

    const lineChannelAccessToken = PropertiesService.getScriptProperties().getProperty("LINE_CHANNEL_ACCESS_TOKEN");
    if (lineChannelAccessToken === null) {
      throw new Error("Gas properties LINE_CHANNEL_ACCESS_TOKEN is not null!");
    }

    const lineGroupId = PropertiesService.getScriptProperties().getProperty("LINE_GROUP_ID");
    if (lineGroupId === null) {
      throw new Error("Gas properties LINE_GROUP_ID is not null!");
    }

    const masterSpreadSheetFile = PropertiesService.getScriptProperties().getProperty("MASTER_SPREAD_SHEET_FILE");
    if (masterSpreadSheetFile === null) {
      throw new Error("Gas properties MASTER_SPREAD_SHEET_FILE is not null!");
    }

    const messageDate = PropertiesService.getScriptProperties().getProperty("MESSAGE_DATE");
    if (messageDate === null) {
      throw new Error("Gas properties MESSAGE_DATE is not null!");
    }

    this.DISPLAY_HIMSELF = displayHimself;
    this.DISPLAY_FAMILY = displayFamily;
    this.DRIVE_FOLDER_NAME = driveFolderName;
    this.LINE_CHANNEL_ACCESS_TOKEN = lineChannelAccessToken;
    this.LINE_GROUP_ID = lineGroupId;
    this.MASTER_SPREAD_SHEET_FILE = masterSpreadSheetFile;
    this.MESSAGE_DATE = messageDate;
  }

  setMessageDate(messageDate: string) {
    PropertiesService.getScriptProperties().setProperty("MESSAGE_DATE", String(messageDate));
  }
}
