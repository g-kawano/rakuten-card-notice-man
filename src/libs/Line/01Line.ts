const LINE_CHANNEL_ACCESS_TOKEN = PropertiesService.getScriptProperties().getProperty("LINE_CHANNEL_ACCESS_TOKEN");
const LINE_GROUP_ID = PropertiesService.getScriptProperties().getProperty("LINE_GROUP_ID");

// @ts-ignore ts(2304) GAS のライブラリから読み込んでいるため
//https://github.com/kobanyan/line-bot-sdk-gas
const lineClient = new LineBotSDK.Client({
  channelAccessToken: LINE_CHANNEL_ACCESS_TOKEN,
});

/**
 * LINE 操作用クラス
 */
export class Line {
  /**
   * LINE にメッセージをプッシュする
   * push は無料プランでは 200 通までなので使いすぎ注意
   * @param message 送信メッセージ内容
   */
  pushMessage(message: any): void {
    lineClient.pushMessage(LINE_GROUP_ID, message);
  }
}
