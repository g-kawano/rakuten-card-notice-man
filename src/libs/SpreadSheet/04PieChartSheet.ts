import { SpreadSheet } from "@/libs/SpreadSheet/01SpreadSheet";
import { Drive } from "@/libs/Drive/01Drive";
import { Setting } from "@/00Setting";

const pieChartSheetSettings = new Setting();

/**
 * 円グラフシート操作用クラス
 */
export class PieChartSheet extends SpreadSheet {
  sheetName: string;
  constructor(fileName: string, sheetName: string) {
    super(fileName, sheetName);
    this.sheetName = sheetName;
  }

  /**
   * スプレッドシートのチャートの画像データを取得する
   */
  private getChartImg(): GoogleAppsScript.Base.Blob {
    const graph = this.sheet.getCharts();
    return graph[0].getBlob();
  }

  /**
   * スプレッドシートのチャート画像を ドライブにアップロードする
   * @param fileName アップロードするファイル名
   */
  uploadChart(fileName: string): void {
    const drive = new Drive();
    const graphImg = this.getChartImg();
    const folder = drive.findFolderByName(pieChartSheetSettings.DRIVE_FOLDER_NAME);
    folder.createFile(graphImg.setName(fileName));
  }

  /**
   * チャートの画像URLをダウンロードする
   */
  downloadChartUrl(fileName: string): string {
    const drive = new Drive();
    const file = drive.findFileByName(fileName);

    if (file === null) {
      throw new Error("File does not exist!");
    }
    file.setSharing(DriveApp.Access.ANYONE, DriveApp.Permission.VIEW);

    return file.getDownloadUrl();
  }
}
