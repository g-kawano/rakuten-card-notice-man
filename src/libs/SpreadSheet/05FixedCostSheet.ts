import { SpreadSheet } from "@/libs/SpreadSheet/01SpreadSheet";

/**
 * 固定費シート操作用クラス
 */
export class FixedCostSheet extends SpreadSheet {
  sheetName: string;
  constructor(fileName: string, sheetName: string) {
    super(fileName, sheetName);
    this.sheetName = sheetName;
  }

  /**
   * ヘッダーとID を除いた全レコードデータを取得する
   */
  scanRecord(): (string | number)[][] {
    const lastRow = this.sheet.getLastRow();

    // 固定日名と金額の列を取得
    const range = this.sheet.getRange(2, 2, lastRow - 1, 2);
    return range.getValues();
  }
}
