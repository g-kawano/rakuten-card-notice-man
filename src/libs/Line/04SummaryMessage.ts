import { BoxContent, TextContent, ImageContent, SeparatorContent, FillerContent } from "@/libs/Line/02LineMessage";
import { PaymentHistorySheet } from "../SpreadSheet/02PaymentHistorySheet";
import { FixedCostSheet } from "../SpreadSheet/05FixedCostSheet";
import { PieChartSheet } from "../SpreadSheet/04PieChartSheet";
import { FlexBox, Message } from "@line/bot-sdk";

/**
 * サマリーメッセージ用クラス
 */
export class SummaryMessage {
  targetYear: string;
  targetMonth: string;
  fixedCostSheet: FixedCostSheet;
  paymentHistorySheet: PaymentHistorySheet;
  previousPaymentHistorySheet: PaymentHistorySheet;
  pieChartSheet: PieChartSheet;

  constructor(
    targetYear: string,
    targetMonth: string,
    fixedCostSheet: FixedCostSheet,
    paymentHistorySheet: PaymentHistorySheet,
    previousPaymentHistorySheet: PaymentHistorySheet,
    pieChartSheet: PieChartSheet
  ) {
    this.targetYear = targetYear;
    this.targetMonth = targetMonth;
    this.fixedCostSheet = fixedCostSheet;
    this.paymentHistorySheet = paymentHistorySheet;
    this.previousPaymentHistorySheet = previousPaymentHistorySheet;
    this.pieChartSheet = pieChartSheet;
  }

  /**
   * 送信メッセージを返す
   */
  buildSendMessage(): Message {
    return {
      type: "flex",
      altText: "サマリーメッセージ",
      contents: {
        type: "bubble",
        header: this.buildHeaderContent(),
        body: this.buildBodyContent()
      }
    };
  }

  /**
   * flex メッセージのヘッダーの Box コンテントを作成して返す
   */
  buildHeaderContent(): FlexBox {
    const header = new BoxContent({ layout: "vertical", backgroundColor: "#1E90FF" });

    const headerContent = new TextContent({
      text: `${this.targetMonth} 月のまとめ💸`,
      align: "center",
      color: "#FFFFFFFF",
      weight: "bold",
      size: "xl"
    });

    header.addContent(headerContent.textContent);

    return header.boxContent;
  }

  /**
   * flex メッセージのボディの Box コンテントを作成して返す
   */
  buildBodyContent(): FlexBox {
    const bodyContent = new BoxContent({ layout: "vertical" });

    bodyContent.addContent(this.buildSpendingContent());
    bodyContent.addContent(new SeparatorContent("md"));
    bodyContent.addContent(this.buildPreviousMonthAmountComparison());
    bodyContent.addContent(this.buildFixedCostContent());
    bodyContent.addContent(new SeparatorContent("xl"));
    bodyContent.addContent(this.buildPieChartImageContent());

    return bodyContent.boxContent;
  }

  /**
   * ボディの収支表示用の Box コンテントを作成して返す
   */
  buildSpendingContent(): FlexBox {
    const spendingContent = new BoxContent({ layout: "horizontal" });

    const spendingKey = new TextContent({
      text: "支出：",
      weight: "bold",
      size: "md",
      flex: 4
    });

    const totalAmount = this.paymentHistorySheet.sumColumn(2, 4) + this.fixedCostSheet.sumColumn(2, 3);

    const spendingValue = new TextContent({
      text: `¥ ${totalAmount.toLocaleString()}`,
      align: "end",
      flex: 5,
      size: "md",
      weight: "bold",
      decoration: "underline"
    });

    spendingContent.addContent(spendingKey.textContent);
    spendingContent.addContent(spendingValue.textContent);

    return spendingContent.boxContent;
  }

  /**
   * 前月比を計算して返す
   * 当月の合計金額 - 前月の合計金額
   */
  calcPreviousMonthAmountComparison(): number {
    return this.paymentHistorySheet.sumColumn(2, 4) - this.previousPaymentHistorySheet.sumColumn(2, 4);
  }

  /**
   * ボディの前月比の Box コンテントを作成して返す
   */
  buildPreviousMonthAmountComparison(): FlexBox {
    const amountComparisonContent = new BoxContent({ layout: "horizontal", margin: "md" });

    const filler = new FillerContent();

    const momKey = new TextContent({
      text: "前月比",
      weight: "bold",
      size: "md",
      flex: 4
    });

    const amountComparison = this.calcPreviousMonthAmountComparison();

    const allow = amountComparison > 0 ? "↑" : "↓";
    const color = amountComparison > 0 ? "#FF0000" : "#1E90FF";

    const momValue = new TextContent({
      text: `${allow} ¥ ${amountComparison.toLocaleString()}`,
      align: "end",
      flex: 5,
      size: "md",
      weight: "bold",
      color: color
    });

    amountComparisonContent.addContent(filler);
    amountComparisonContent.addContent(momKey.textContent);
    amountComparisonContent.addContent(momValue.textContent);

    return amountComparisonContent.boxContent;
  }

  /**
   * 1 レコード分の固定費を表示する Box コンテントを作成して返す
   * @param isFirstRecord 最初のレコードかどうか
   * @param costName 固定費名
   * @param costValue 固定費金額
   */
  buildFixedCostContentRecord(isFirstRecord: boolean, costName: string, costValue: number): FlexBox {
    const fixedCostRecordContent = new BoxContent({
      layout: "horizontal",
      margin: "xs",
      justifyContent: "flex-start"
    });

    if (isFirstRecord) {
      const filler = new FillerContent();
      const fixedKeyContent = new TextContent({
        flex: 3,
        text: "固定費",
        size: "sm"
      });
      fixedCostRecordContent.addContent(filler);
      fixedCostRecordContent.addContent(fixedKeyContent.textContent);
    } else {
      const filler = new FillerContent(4);
      fixedCostRecordContent.addContent(filler);
    }

    const nameContent = new TextContent({
      text: costName,
      flex: 3,
      size: "sm"
    });

    const valueContent = new TextContent({
      text: `¥ ${costValue.toLocaleString()}`,
      flex: 3,
      align: "end"
    });

    fixedCostRecordContent.addContent(nameContent.textContent);
    fixedCostRecordContent.addContent(valueContent.textContent);

    return fixedCostRecordContent.boxContent;
  }

  /**
   * ボディの固定費の Box コンテントを作成して返す
   */
  buildFixedCostContent(): FlexBox {
    const amountComparisonContent = new BoxContent({ layout: "vertical", margin: "sm" });

    for (const [index, record] of this.fixedCostSheet.scanRecord().entries()) {
      const isFirst = index === 0;

      const costName = record[0];
      const costValue = record[1];

      if (typeof costName !== "string") {
        throw new Error("Cost name is not string!");
      }

      if (typeof costValue !== "number") {
        throw new Error("Cost value is not number!");
      }

      amountComparisonContent.addContent(this.buildFixedCostContentRecord(isFirst, costName, costValue));
    }

    return amountComparisonContent.boxContent;
  }

  /**
   * ボディの円グラフの Image コンテントを作成して返す
   */
  buildPieChartImageContent(): ImageContent {
    const chartFileName = `PieChart-${this.targetMonth}`;
    this.pieChartSheet.uploadChart(chartFileName);
    const imageUrl = this.pieChartSheet.downloadChartUrl(chartFileName);

    const pieChartImageContent = new ImageContent({
      url: imageUrl,
      size: "full",
      align: "center"
    });

    return pieChartImageContent;
  }
}
