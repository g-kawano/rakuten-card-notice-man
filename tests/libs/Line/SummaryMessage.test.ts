import { SummaryMessage } from "@/libs/Line/04SummaryMessage";
import { Setting } from "@/00Setting";
import { FixedCostSheet } from "@/libs/SpreadSheet/05FixedCostSheet";
import { PaymentHistorySheet } from "@/libs/SpreadSheet/02PaymentHistorySheet";
import { FlexBox } from "@line/bot-sdk";
import { PieChartSheet } from "@/libs/SpreadSheet/04PieChartSheet";

jest.mock("@/00Setting");
jest.mock("@/libs/SpreadSheet/01SpreadSheet");
jest.mock("@/libs/SpreadSheet/02PaymentHistorySheet");
jest.mock("@/libs/SpreadSheet/05FixedCostSheet");
jest.mock("@/libs/SpreadSheet/04PieChartSheet");

describe("SummaryMessage", () => {
  let summaryMessage: SummaryMessage;
  let setting: Setting;
  let paymentHistorySheet: PaymentHistorySheet;
  let previousPaymentHistorySheet: PaymentHistorySheet;
  let fixedCostSheet: FixedCostSheet;
  let pieChartSheet: PieChartSheet;

  beforeEach(() => {
    const targetYear = "2023";
    const targetMonth = "5";

    setting = new Setting() as jest.Mocked<Setting>;
    paymentHistorySheet = new PaymentHistorySheet(
      "dummyFileName",
      "dummySheetName"
    ) as jest.Mocked<PaymentHistorySheet>;

    previousPaymentHistorySheet = new PaymentHistorySheet(
      "dummyFileName",
      "dummySheetName"
    ) as jest.Mocked<PaymentHistorySheet>;

    fixedCostSheet = new FixedCostSheet("dummyFileName", "dummySheetName") as jest.Mocked<FixedCostSheet>;

    pieChartSheet = new PieChartSheet("dummyFileName", "dummySheetName") as jest.Mocked<PieChartSheet>;

    setting.MASTER_SPREAD_SHEET_FILE = "dummy";

    summaryMessage = new SummaryMessage(
      targetYear,
      targetMonth,
      fixedCostSheet,
      paymentHistorySheet,
      previousPaymentHistorySheet,
      pieChartSheet
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe("constructor", () => {
    it("should initialize the instance correctly", () => {
      expect(summaryMessage).toBeInstanceOf(SummaryMessage);
      expect(summaryMessage.targetYear).toBe("2023");
      expect(summaryMessage.targetMonth).toBe("5");
      expect(summaryMessage.fixedCostSheet).toBeInstanceOf(FixedCostSheet);
      expect(summaryMessage.paymentHistorySheet).toBeInstanceOf(PaymentHistorySheet);
      expect(summaryMessage.previousPaymentHistorySheet).toBeInstanceOf(PaymentHistorySheet);
      expect(summaryMessage.pieChartSheet).toBeInstanceOf(PieChartSheet);
    });
  });

  describe("buildSendMessage", () => {
    it("should return correct send message object", () => {
      jest.spyOn(summaryMessage, "buildHeaderContent").mockReturnValue("headerMock" as unknown as FlexBox);
      jest.spyOn(summaryMessage, "buildBodyContent").mockReturnValue("bodyMock" as unknown as FlexBox);

      const result = summaryMessage.buildSendMessage();

      expect(result).toMatchObject({
        type: "flex",
        altText: "サマリーメッセージ",
        contents: {
          type: "bubble",
          header: "headerMock",
          body: "bodyMock"
        }
      });
    });
  });

  describe("buildHeaderContent", () => {
    it("should return correct header object", () => {
      const result = summaryMessage.buildHeaderContent();

      expect(result).toMatchSnapshot();
    });
  });

  describe("buildBodyContent", () => {
    it("should return correct body object", () => {
      paymentHistorySheet.sumColumn = jest.fn().mockReturnValue(2000);
      previousPaymentHistorySheet.sumColumn = jest.fn().mockReturnValue(2000);
      fixedCostSheet.sumColumn = jest.fn().mockReturnValue(2000);
      fixedCostSheet.scanRecord = jest.fn().mockReturnValue([
        ["固定費1", 1000],
        ["固定費2", 200],
        ["固定費3", 0]
      ]);
      pieChartSheet.uploadChart = jest.fn();
      pieChartSheet.downloadChartUrl = jest.fn().mockReturnValue("https://dummy.com");

      const result = summaryMessage.buildBodyContent();

      expect(result).toMatchSnapshot();
    });
  });

  describe("buildSpendingContent", () => {
    it("should return correct spending content object", () => {
      paymentHistorySheet.sumColumn = jest.fn().mockReturnValue(2000);
      previousPaymentHistorySheet.sumColumn = jest.fn().mockReturnValue(2000);
      fixedCostSheet.sumColumn = jest.fn().mockReturnValue(2000);

      const result = summaryMessage.buildSpendingContent();

      expect(result).toHaveProperty("type", "box");
      expect(result).toHaveProperty("layout", "horizontal");

      expect(result.contents[0]).toHaveProperty("type", "text");
      expect(result.contents[0]).toHaveProperty("text", "支出：");
      expect(result.contents[0]).toHaveProperty("wrap", true);
      expect(result.contents[0]).toHaveProperty("size", "md");
      expect(result.contents[0]).toHaveProperty("flex", 4);
      expect(result.contents[0]).toHaveProperty("weight", "bold");

      expect(result.contents[1]).toHaveProperty("type", "text");
      expect(result.contents[1]).toHaveProperty("text", "¥ 4,000");
      expect(result.contents[1]).toHaveProperty("wrap", true);
      expect(result.contents[1]).toHaveProperty("size", "md");
      expect(result.contents[1]).toHaveProperty("flex", 5);
      expect(result.contents[1]).toHaveProperty("weight", "bold");
      expect(result.contents[1]).toHaveProperty("align", "end");
    });
  });

  describe("calcPreviousMonthAmountComparison", () => {
    it("should return correct amount comparison positive value", () => {
      jest.spyOn(paymentHistorySheet, "sumColumn").mockReturnValue(10000);
      jest.spyOn(previousPaymentHistorySheet, "sumColumn").mockReturnValue(2500);

      const result = summaryMessage.calcPreviousMonthAmountComparison();

      expect(result).toBe(7500);
    });

    it("should return correct amount comparison negative value", () => {
      jest.spyOn(paymentHistorySheet, "sumColumn").mockReturnValue(2500);
      jest.spyOn(previousPaymentHistorySheet, "sumColumn").mockReturnValue(10000);

      const result = summaryMessage.calcPreviousMonthAmountComparison();

      expect(result).toBe(-7500);
    });
  });

  describe("buildPreviousMonthAmountComparison", () => {
    it("should return correct amount comparison content when calc result positive value", () => {
      jest.spyOn(paymentHistorySheet, "sumColumn").mockReturnValue(10000);
      jest.spyOn(previousPaymentHistorySheet, "sumColumn").mockReturnValue(2500);

      const result = summaryMessage.buildPreviousMonthAmountComparison();

      expect(result).toHaveProperty("type", "box");
      expect(result).toHaveProperty("layout", "horizontal");
      expect(result).toHaveProperty("margin", "md");

      expect(result.contents[0]).toHaveProperty("type", "filler");

      expect(result.contents[1]).toHaveProperty("type", "text");
      expect(result.contents[1]).toHaveProperty("text", "前月比");
      expect(result.contents[1]).toHaveProperty("wrap", true);
      expect(result.contents[1]).toHaveProperty("flex", 4);
      expect(result.contents[1]).toHaveProperty("weight", "bold");
      expect(result.contents[1]).toHaveProperty("size", "md");

      expect(result.contents[2]).toHaveProperty("type", "text");
      expect(result.contents[2]).toHaveProperty("text", "↑ ¥ 7,500");
      expect(result.contents[2]).toHaveProperty("wrap", true);
      expect(result.contents[2]).toHaveProperty("flex", 5);
      expect(result.contents[2]).toHaveProperty("weight", "bold");
      expect(result.contents[2]).toHaveProperty("size", "md");
      expect(result.contents[2]).toHaveProperty("align", "end");
      expect(result.contents[2]).toHaveProperty("color", "#FF0000");
    });

    it("should return correct amount comparison content when calc result negative value", () => {
      jest.spyOn(paymentHistorySheet, "sumColumn").mockReturnValue(2500);
      jest.spyOn(previousPaymentHistorySheet, "sumColumn").mockReturnValue(10000);

      const result = summaryMessage.buildPreviousMonthAmountComparison();

      expect(result).toHaveProperty("type", "box");
      expect(result).toHaveProperty("layout", "horizontal");
      expect(result).toHaveProperty("margin", "md");

      expect(result.contents[0]).toHaveProperty("type", "filler");

      expect(result.contents[1]).toHaveProperty("type", "text");
      expect(result.contents[1]).toHaveProperty("text", "前月比");
      expect(result.contents[1]).toHaveProperty("wrap", true);
      expect(result.contents[1]).toHaveProperty("flex", 4);
      expect(result.contents[1]).toHaveProperty("weight", "bold");
      expect(result.contents[1]).toHaveProperty("size", "md");

      expect(result.contents[2]).toHaveProperty("type", "text");
      expect(result.contents[2]).toHaveProperty("text", "↓ ¥ -7,500");
      expect(result.contents[2]).toHaveProperty("wrap", true);
      expect(result.contents[2]).toHaveProperty("flex", 5);
      expect(result.contents[2]).toHaveProperty("weight", "bold");
      expect(result.contents[2]).toHaveProperty("size", "md");
      expect(result.contents[2]).toHaveProperty("align", "end");
      expect(result.contents[2]).toHaveProperty("color", "#1E90FF");
    });
  });

  describe("buildFixedCostContentRecord", () => {
    it("should return correct fixed content when first record is true", () => {
      fixedCostSheet.sumColumn = jest.fn().mockReturnValue(2000);
      fixedCostSheet.scanRecord = jest.fn().mockReturnValue([
        ["固定費1", 1000],
        ["固定費2", 200],
        ["固定費3", 0]
      ]);

      const result = summaryMessage.buildFixedCostContentRecord(true, "cost", 1000);

      expect(result).toHaveProperty("type", "box");
      expect(result).toHaveProperty("layout", "horizontal");
      expect(result).toHaveProperty("margin", "xs");
      expect(result).toHaveProperty("justifyContent", "flex-start");

      expect(result.contents[0]).toHaveProperty("type", "filler");

      expect(result.contents[1]).toHaveProperty("type", "text");
      expect(result.contents[1]).toHaveProperty("text", "固定費");
      expect(result.contents[1]).toHaveProperty("size", "sm");

      expect(result.contents[2]).toHaveProperty("type", "text");
      expect(result.contents[2]).toHaveProperty("text", "cost");
      expect(result.contents[2]).toHaveProperty("size", "sm");
      expect(result.contents[2]).toHaveProperty("flex", 3);

      expect(result.contents[3]).toHaveProperty("type", "text");
      expect(result.contents[3]).toHaveProperty("text", "¥ 1,000");
      expect(result.contents[3]).toHaveProperty("flex", 3);
      expect(result.contents[3]).toHaveProperty("align", "end");
    });

    it("should return correct fixed content when first record is false", () => {
      fixedCostSheet.sumColumn = jest.fn().mockReturnValue(2000);
      fixedCostSheet.scanRecord = jest.fn().mockReturnValue([
        ["固定費1", 1000],
        ["固定費2", 200],
        ["固定費3", 0]
      ]);

      const result = summaryMessage.buildFixedCostContentRecord(false, "cost", 1000);

      expect(result).toHaveProperty("type", "box");
      expect(result).toHaveProperty("layout", "horizontal");
      expect(result).toHaveProperty("margin", "xs");
      expect(result).toHaveProperty("justifyContent", "flex-start");

      expect(result.contents[0]).toHaveProperty("type", "filler");
      expect(result.contents[0]).toHaveProperty("flex", 4);

      expect(result.contents[1]).toHaveProperty("type", "text");
      expect(result.contents[1]).toHaveProperty("text", "cost");
      expect(result.contents[1]).toHaveProperty("size", "sm");
      expect(result.contents[1]).toHaveProperty("flex", 3);

      expect(result.contents[2]).toHaveProperty("type", "text");
      expect(result.contents[2]).toHaveProperty("text", "¥ 1,000");
      expect(result.contents[2]).toHaveProperty("flex", 3);
      expect(result.contents[2]).toHaveProperty("align", "end");
    });
  });

  describe("buildFixedCostContent", () => {
    it("should return correct fixed content", () => {
      fixedCostSheet.scanRecord = jest.fn().mockReturnValue([
        ["固定費1", 1000],
        ["固定費2", 200],
        ["固定費3", 0]
      ]);

      const mockBuildFixedCostContentRecord = jest
        .spyOn(summaryMessage, "buildFixedCostContentRecord")
        .mockReturnValue("mock" as unknown as FlexBox);

      const result = summaryMessage.buildFixedCostContent();

      expect(result).toHaveProperty("type", "box");
      expect(result).toHaveProperty("layout", "vertical");
      expect(result).toHaveProperty("margin", "sm");

      expect(mockBuildFixedCostContentRecord).toHaveBeenCalledTimes(3);

      expect(mockBuildFixedCostContentRecord).toHaveBeenNthCalledWith(1, true, "固定費1", 1000);
      expect(mockBuildFixedCostContentRecord).toHaveBeenNthCalledWith(2, false, "固定費2", 200);
      expect(mockBuildFixedCostContentRecord).toHaveBeenNthCalledWith(3, false, "固定費3", 0);

      expect(result.contents[0]).toEqual("mock");
    });

    it("should throw error when cost name is not string", () => {
      jest.spyOn(fixedCostSheet, "scanRecord").mockReturnValue([[1000, 1000]]);

      expect(() => summaryMessage.buildFixedCostContent()).toThrow(Error("Cost name is not string!"));
    });

    it("should throw error when cost value is not number", () => {
      jest.spyOn(fixedCostSheet, "scanRecord").mockReturnValue([["固定費1", "固定費1"]]);

      expect(() => summaryMessage.buildFixedCostContent()).toThrow(Error("Cost value is not number!"));
    });
  });

  describe("buildPieChartImageContent", () => {
    it("should return correct pie chart image content", () => {
      pieChartSheet.uploadChart = jest.fn();
      pieChartSheet.downloadChartUrl = jest.fn().mockReturnValue("https://dummy.com");

      const result = summaryMessage.buildPieChartImageContent();

      expect(result).toHaveProperty("type", "image");
      expect(result).toHaveProperty("url", "https://dummy.com");
      expect(result).toHaveProperty("size", "full");
      expect(result).toHaveProperty("align", "center");
    });
  });
});
