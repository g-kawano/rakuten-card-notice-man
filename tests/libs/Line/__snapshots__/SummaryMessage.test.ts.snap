// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryMessage buildBodyContent should return correct body object 1`] = `
{
  "backgroundColor": undefined,
  "contents": [
    {
      "backgroundColor": undefined,
      "contents": [
        {
          "align": undefined,
          "color": undefined,
          "decoration": undefined,
          "flex": 4,
          "margin": undefined,
          "size": "md",
          "text": "ÊîØÂá∫Ôºö",
          "type": "text",
          "weight": "bold",
          "wrap": true,
        },
        {
          "align": "end",
          "color": undefined,
          "decoration": "underline",
          "flex": 5,
          "margin": undefined,
          "size": "md",
          "text": "¬• 4,000",
          "type": "text",
          "weight": "bold",
          "wrap": true,
        },
      ],
      "justifyContent": undefined,
      "layout": "horizontal",
      "margin": undefined,
      "type": "box",
    },
    SeparatorContent {
      "margin": "md",
      "type": "separator",
    },
    {
      "backgroundColor": undefined,
      "contents": [
        FillerContent {
          "flex": undefined,
          "type": "filler",
        },
        {
          "align": undefined,
          "color": undefined,
          "decoration": undefined,
          "flex": 4,
          "margin": undefined,
          "size": "md",
          "text": "ÂâçÊúàÊØî",
          "type": "text",
          "weight": "bold",
          "wrap": true,
        },
        {
          "align": "end",
          "color": "#1E90FF",
          "decoration": undefined,
          "flex": 5,
          "margin": undefined,
          "size": "md",
          "text": "‚Üì ¬• 0",
          "type": "text",
          "weight": "bold",
          "wrap": true,
        },
      ],
      "justifyContent": undefined,
      "layout": "horizontal",
      "margin": "md",
      "type": "box",
    },
    {
      "backgroundColor": undefined,
      "contents": [
        {
          "backgroundColor": undefined,
          "contents": [
            FillerContent {
              "flex": undefined,
              "type": "filler",
            },
            {
              "align": undefined,
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": "sm",
              "text": "Âõ∫ÂÆöË≤ª",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
            {
              "align": undefined,
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": "sm",
              "text": "Âõ∫ÂÆöË≤ª1",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
            {
              "align": "end",
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": undefined,
              "text": "¬• 1,000",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
          ],
          "justifyContent": "flex-start",
          "layout": "horizontal",
          "margin": "xs",
          "type": "box",
        },
        {
          "backgroundColor": undefined,
          "contents": [
            FillerContent {
              "flex": 4,
              "type": "filler",
            },
            {
              "align": undefined,
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": "sm",
              "text": "Âõ∫ÂÆöË≤ª2",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
            {
              "align": "end",
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": undefined,
              "text": "¬• 200",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
          ],
          "justifyContent": "flex-start",
          "layout": "horizontal",
          "margin": "xs",
          "type": "box",
        },
        {
          "backgroundColor": undefined,
          "contents": [
            FillerContent {
              "flex": 4,
              "type": "filler",
            },
            {
              "align": undefined,
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": "sm",
              "text": "Âõ∫ÂÆöË≤ª3",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
            {
              "align": "end",
              "color": undefined,
              "decoration": undefined,
              "flex": 3,
              "margin": undefined,
              "size": undefined,
              "text": "¬• 0",
              "type": "text",
              "weight": undefined,
              "wrap": true,
            },
          ],
          "justifyContent": "flex-start",
          "layout": "horizontal",
          "margin": "xs",
          "type": "box",
        },
      ],
      "justifyContent": undefined,
      "layout": "vertical",
      "margin": "sm",
      "type": "box",
    },
    SeparatorContent {
      "margin": "xl",
      "type": "separator",
    },
    ImageContent {
      "align": "center",
      "margin": undefined,
      "size": "full",
      "type": "image",
      "url": "https://dummy.com",
    },
  ],
  "justifyContent": undefined,
  "layout": "vertical",
  "margin": undefined,
  "type": "box",
}
`;

exports[`SummaryMessage buildHeaderContent should return correct header object 1`] = `
{
  "backgroundColor": "#1E90FF",
  "contents": [
    {
      "align": "center",
      "color": "#FFFFFFFF",
      "decoration": undefined,
      "flex": undefined,
      "margin": undefined,
      "size": "xl",
      "text": "5 Êúà„ÅÆ„Åæ„Å®„ÇÅüí∏",
      "type": "text",
      "weight": "bold",
      "wrap": true,
    },
  ],
  "justifyContent": undefined,
  "layout": "vertical",
  "margin": undefined,
  "type": "box",
}
`;
